packages/schema/bind/src/bindings/assemblyscript/wasm/templates/module-type/module-ts.mustache

```diff
@@ -0,0 +1,43 @@
+import * as Types from "..";
+
+{{> serialization_imports}}
+
+import {
+{{#methods}}
+  deserialize{{name}}Args,
+  serialize{{name}}Result{{^last}},{{/last}}
+{{/methods}}
+} from "./serialization";
+
+import {
+  wrap_invoke_args,
+  wrap_invoke_result,
+  wrap_invoke_error,
+  wrap_load_env,
+  InvokeArgs,
+} from "@polywrap/wasm-as";
+
+export abstract class IModule {
+  {{#hasEnv}}private __env__: Types.Env | null = null;
+
+  public get env(): Types.Env {
```

Comment:
Hey @Niraj-Kamdar @namesty, I think we have a conflict in how we treat `env` here.

First, this getter method is added to the module bindings if the ABI has an `envType` defined (see `moduleHasEnv.ts` and `{{#hasEnv}}` above). It will throw an error if the `env` is not set, stating "it is required by this module".

Now for the inconsistency. We previously added the `@env(required: boolean)` directive so that we can selectively require the environment for specific module methods. For example, see `test-cases/cases/bind/sanity/input/schema.graphql`. This means that by default the `env` is optional for all methods, and is selectively required on a per-method basis.

I think we should take one of the 2 possible paths:
1. Respect this pattern of having `env` optional, and having it required on a per-method basis. This means that we will probably go back to passing the `env` as an argument to the method.
2. Revert the `@env` directive, and go back to requiring the `env` on a per-module basis.

Thoughts?

Reply:
Actually after further inspection @Niraj-Kamdar, the per-method requirement of the environment is already handled well within the `wrap/Module/wrapped.ts` file. See the `objectMethodWrapped` example:
```typescript
export function objectMethodWrapped(module: IModule, argsBuf: ArrayBuffer, env_size: u32): ArrayBuffer {
  if (env_size == 0) {
    throw new Error("Environment is not set, and it is required by method 'objectMethod'")
  }
  
  const envBuf = wrap_load_env(env_size);
  module.__setEnv__(Types.Env.fromBuffer(envBuf));
  ...
}
```

This means that we do not have to handle this scenario within the `IModule` bindings, and we can instead do something like this:
```typescript
export abstract class IModule {
  private __env__: Types.Env | null = null;

  public get env(): Types.Env {
    if (this.__env__ === null) {
      throw new Error("Environment is not set, and is not required by this method.");
    }
    return this.__env__ as Types.Env;
  }

  public __setEnv__(env: Types.Env): void {
    this.__env__ = env;
  }
}
```

It's not the cleanest pattern... I would honestly prefer going back to having the `env` be an argument of the module's method.

Reply:
This is to give the same ergonomics of the plugins to the wrapper. As long as the pattern is consistent I don't mind the way in which we implement it. 

If we prefer method args env then we should do the same in the plugin, if we prefer `this.env` then we should do the same in the wrapper.

Reply:
using env in function params